### DISCUSSION TITLE
Using "where" inside "include" is not working

### DISCUSSION THREAD

**Discussion by ManishKarki1997** [prisma/prisma#11172] [Q&A]
Hi, I ran into some issue. 

As the documentation shows here https://www.prisma.io/docs/concepts/components/prisma-client/filtering-and-sorting
```
...
 include: {
    posts: {
      where: {
        published: true,
      },
    },
  },

```

This is supposed to be working, where query inside "include". But, when I try similar, I get an error 
```

Type '{ where: {}; include: { user: true; }; }' is not assignable to type 'boolean | QnAArgs | undefined'.
  Object literal may only specify known properties, and 'where' does not exist in type 'QnAArgs'

```

My schema is 

```

model QnA {
  id         Int      @id @default(autoincrement())
  qnaText    String
  question   QnA?     @relation("QuestionAnswers", fields: [questionId], references: [id])
  questionId Int?
  @@map("qna")
}

```

My query is

```

const questions = await prisma.qnA.findMany({
        take: parseInt(take),
        skip: parseInt(skip),
        where: {
          product: {
            slug: productSlug,
          },
          questionId: null, //ensures only questions are fetched. answer has a questionId field available
        },
        include: {
          user: true,
          question: {
            //this level should be the answer(if there is) to the above level question
            include: {
              user: true,
            },
          },
        },
      });

```

The idea is to fetch all the questions (which has questionId of null) then include their answers if available. But this query only gives me the questions, even though the there are questions(answer in my case, just the naming is question) whose questionId references the id field.

I found a similar question in stackoverflow by someone else here
https://stackoverflow.com/questions/69731954/filtering-relations-with-include-and-where

Thanks!

**Comment by samrathchadha** [comment]
If you're filtering items that can be fetched via include without custom logic, you can filter out the required items via `array.filter()` but since what you're doing is basically multiple level inclusion, that isn't an implemented feature yet.
You could ask for this in an issue as a feature request though.

**Reply by ManishKarki1997** [reply]
But I see similar thing done in the docs too. using the where clause inside include, like this.

```
 include: {
    posts: {
      where: {
        published: true,
      },
    },
  },

```

**Reply by samrathchadha** [reply]
could you refer to where you've seen this?

**Reply by ManishKarki1997** [reply]
Hi, sure. the link is here
https://www.prisma.io/docs/concepts/components/prisma-client/filtering-and-sorting

Maybe I'm not understanding this properly.

The full code is 
```
const result = await prisma.user.findMany({
  where: {
    email: {
      endsWith: 'prisma.io',
    },
    posts: {
      some: {
        published: true,
      },
    },
  },
  include: {
    posts: {
      where: {
        published: true,
      },
    },
  },
})
```

**Reply by samrathchadha** [reply]
So the difference between what you're doing, and they're doing is: you're trying to include something from within an included element. The example, is filtering an included element.

**Reply by ManishKarki1997** [reply]
Hmm, I don't quite understand what you're saying sorry. Is there any way to make my code work?

**Reply by samrathchadha** [reply]
No, i don't think so

**Reply by ManishKarki1997** [reply]
I see. Last question, Do you have any idea why this is not working? 

```
const questions = await prisma.qnA.findMany({
        take: parseInt(take),
        skip: parseInt(skip),
        where: {
          product: {
            slug: productSlug,
          },
          questionId: null, //ensures only questions are fetched. answer has a questionId field available
        },
        include: {
          user: true,
          question: {
            //this level should be the answer(if there is) to the above level question
            include: {
              user: true,
            },
          },
        },
      });

```



I'm fetching qnA (questions and answers) where the questionId is null (meaning, only questions, not answers). and then, I'm including the "question" field which will contain the answer to that question. As far as I can tell, this should be working. but it returns empty array. If I remove "questionId:null" from the where clause, I get questions and answers both.

**Reply by samrathchadha** [reply]
You're still using `include` nested inside an `include` statement. This won't work

**Reply by ManishKarki1997** [reply]
I don't think that's true because nested includes did work in my other cases. Its just the "where" clause inside "include" is not working

**Reply by samrathchadha** [reply]
Could you share more code(that's working) with your nested include?

**Reply by ManishKarki1997** [reply]
Sure.

```
const user = await prisma.user.findFirst({
        where: {
          email,
        },
        include: {
          wishlist: {
            include: {
              product: {
                include: {
                  images: true,
                  pricing: {
                    take: 1,
                    orderBy: {
                      createdAt: "desc",
                    },
                  },
                  productDiscount: {
                    where: {
                      isValid: true,
                    },
                  },
                },
              },
            },
          },
          cart: {
            include: {
              product: {
                include: {
                  images: true,
                  pricing: {
                    take: 1,
                    orderBy: {
                      createdAt: "desc",
                    },
                  },
                  productDiscount: {
                    where: {
                      isValid: true,
                    },
                  },
                },
              },
            },
          },
        },
      });
```

**Reply by samrathchadha** [reply]
That's pretty strange, I'm not sure how to fix it, you could ask someone more familiar with the prisma ecosystem.

**Reply by NourMawla98** [reply]
@ManishKarki1997 where you able to find a solution for this issue? I am facing the same problem

**Comment by old-zoomer** [comment]
@ManishKarki1997

If I understood your idea correctly you want to have 1 to many self relation between `question`(1) and (many)`answers`
Currently your schema defines only parent relation from `answer` to `question`
You need to add child relation definition from `question` to `answers`

Here's modified schema to achieve this

```diff
model QnA {
  id         Int      @id @default(autoincrement())
  qnaText    String
  question   QnA?     @relation("QuestionAnswers", fields: [questionId], references: [id])
  questionId Int?
+ answers  QnA[]  @relation("QuestionAnswers") // child relation
  @@map("qna")
}
```

Here is documentation for this: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/self-relations#one-to-many-self-relations

I don't believe this will modify actual SQL schema, because in SQL tables only parent relation will be tracked anyway, but I think with this line prisma will add needed JOIN to the SQL query

Also if `answers` should not have further child `answers`/`questions` I would highly recommend to define separate model for answers, then you schema would look like this:

```
model Question {
  id         Int      @id @default(autoincrement())
  questionText    String
  answers  Answer[] 
  @@map("question")
}

model Answer {
  id         Int      @id @default(autoincrement())
  answerText    String
  question  Question @relation(fields: [questionId], references: [id])
  questionId Int
  @@map("answer")
}
```
Here is documentation for this: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/one-to-many-relations

1. In this case there is no way to add child items to answer. So if this is your case it's better to restrict yourself on schema level.
2. I think this schema is more readable

@NourMawla98

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
